File: ios/Ilaria/PhoneLocal.swift
import Foundation

struct ChatOut {
    var text: String
    var phone: [[String: Any]] = []
    var fromPc: Bool = false
}

/// Offline / local command router for iPhone (parity with Android PhoneLocal).
enum PhoneLocal {
    static func handle(raw: String, prefs: Prefs) -> ChatOut? {
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if let r = try? NSRegularExpression(pattern: #"^(?:hey\s+)?ilaria\b[\s,.:\-]*"#, options: .caseInsensitive) {
            text = r.stringByReplacingMatches(in: text, range: NSRange(text.startIndex..., in: text), withTemplate: "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if text.isEmpty { return ChatOut(text: "Te escucho.") }
        let lower = text.lowercased()

        if lower.range(of: #"\b(ayuda|help|comandos)\b"#, options: .regularExpression) != nil {
            return ChatOut(text: "En iPhone: apps (no bancarias), Spotify/YouTube, linterna, mapas, llamadas, SMS/WhatsApp borrador, portapapeles, traducir, notas. Enlazá la PC por Wi‑Fi para el cerebro completo.")
        }
        if lower.range(of: #"\b(hora|fecha)\b"#, options: .regularExpression) != nil || ["ahora", "hora", "fecha"].contains(lower) {
            let f = DateFormatter()
            f.locale = Locale(identifier: "es_AR")
            f.dateFormat = "EEEE d 'de' MMMM, HH:mm"
            return ChatOut(text: f.string(from: Date()))
        }
        if lower.range(of: #"\b(linterna|torch)\b"#, options: .regularExpression) != nil {
            let off = lower.range(of: #"\b(apag|off)\b"#, options: .regularExpression) != nil
            return hands("torch", target: off ? "off" : "on", reply: "Listo.")
        }
        if lower.contains("camara") || lower.contains("cámara") {
            return hands("camera", reply: "Cámara.")
        }
        if lower.range(of: #"\b(wifi|wi-?fi)\b"#, options: .regularExpression) != nil {
            return hands("wifi", reply: "Wi‑Fi.")
        }
        if lower.contains("bluetooth") { return hands("bluetooth", reply: "Bluetooth.") }
        if lower.range(of: #"\b(ajustes|settings)\b"#, options: .regularExpression) != nil {
            return hands("settings", reply: "Ajustes.")
        }
        if lower.range(of: #"\b(silenci|mute)\b"#, options: .regularExpression) != nil {
            return hands("volume", target: "mute", reply: "Abrí Sonidos en Ajustes (iOS no deja mute silencioso a apps).")
        }
        if let m = lower.range(of: #"volumen\s+(\d{1,3})"#, options: .regularExpression) {
            let num = String(lower[m]).components(separatedBy: CharacterSet.decimalDigits.inverted).joined()
            return hands("volume", target: num, reply: "Volumen: usá los botones laterales o Ajustes.")
        }
        if lower.contains("captura") || lower.contains("screenshot") {
            return hands("screenshot", reply: "Captura: botón lateral + volumen arriba.")
        }
        if lower.contains("traduc") {
            let q = text.replacingOccurrences(of: #"(?i)^(?:traduc[ií]|traducir|translate)\s+"#, with: "", options: .regularExpression)
            return hands("translate", text: q, reply: "Traduzco.")
        }
        if lower.hasPrefix("copia ") || lower.hasPrefix("copiá ") || lower.hasPrefix("copiar ") {
            let body = text.replacingOccurrences(of: #"(?i)^(?:copi[aá]|copiar)\s*[:\-]?\s*"#, with: "", options: .regularExpression)
            return hands("clipboard", text: body, reply: "Copiado.")
        }
        if lower.contains("portapapeles") {
            return hands("clipboard_get", reply: "Leo el portapapeles.")
        }
        if let call = match(#"(?i)(?:llam[aá]|marca)\s+(?:al\s+|a\s+)?([+\d][\d\s\-()]{6,})"#, text) {
            let digits = call.replacingOccurrences(of: #"[^\d+]"#, with: "", options: .regularExpression)
            return hands("call", target: digits, reply: "Abro el marcador.")
        }
        if let wa = match(#"(?i)(?:whats?app|wpp)\s+(?:a|al)?\s*([+\d][\d\s\-]{7,18})\s*[:\-]\s*(.+)$"#, text) {
            let parts = wa.split(separator: "|", maxSplits: 1).map(String.init)
            if parts.count == 2 {
                return hands("whatsapp", target: parts[0].filter(\.isNumber), text: parts[1], reply: "WhatsApp borrador.")
            }
        }
        if let maps = match(#"(?i)^(?:c[oó]mo\s+llego(?:\s+a)?|mapas?|ruta(?:\s+a)?|ll[eé]vame\s+a|naveg[aá]\s+(?:a|hacia))\s+(.+)$"#, text) {
            return hands("navigate", target: maps, reply: "Te armo la ruta.")
        }
        if lower.hasPrefix("anotá ") || lower.hasPrefix("anota ") || lower.hasPrefix("nota:") {
            let body = text.replacingOccurrences(of: #"(?i)^(?:anot[aá]|nota[:\s]+)\s*"#, with: "", options: .regularExpression)
            prefs.addNote(body)
            return ChatOut(text: "Nota guardada en el iPhone.")
        }
        if ["notas", "mis notas"].contains(lower) {
            let lines = prefs.notes()
            return ChatOut(text: lines.isEmpty ? "No hay notas." : lines.enumerated().map { "\($0.offset + 1). \($0.element)" }.joined(separator: "\n"))
        }
        if text.lowercased().hasPrefix("http") {
            return hands("browser", target: text, reply: "Abro el link.")
        }
        if let opened = match(#"(?i)(?:abr[ií]|abrime|abrir|abre|open)\s+(?:la\s+|el\s+)?(.+)$"#, text) {
            if BankApps.blocked(opened) { return ChatOut(text: "No abro apps bancarias.") }
            let t = opened.lowercased()
            if t == "youtube" || t == "yt" {
                return hands("browser", target: "https://www.youtube.com", reply: "YouTube.")
            }
            if t == "gmail" || t == "correo" {
                return hands("browser", target: "https://mail.google.com", reply: "Gmail.")
            }
            return hands("open_app", target: opened, reply: "Abro \(opened).")
        }
        let bare = ["spotify", "whatsapp", "telegram", "instagram", "youtube", "chrome", "maps", "netflix"]
        if bare.contains(lower) {
            return hands("open_app", target: lower, reply: "Abro \(lower).")
        }
        return nil
    }

    static func fallback(prefs: Prefs, linked: Bool) -> String {
        let who = prefs.displayName.isEmpty ? (prefs.username.isEmpty ? "vos" : prefs.username) : prefs.displayName
        if linked { return "No pude completar eso en el iPhone. Probá de nuevo." }
        return "Estoy en el iPhone, \(who) — independiente de la PC. Apps, notas y linterna van acá. Si querés, en Perfil sincronizás con la PC (mismo Wi‑Fi)."
    }

    private static func hands(_ action: String, target: String = "", text: String = "", reply: String) -> ChatOut {
        var obj: [String: Any] = ["action": action]
        if !target.isEmpty { obj["target"] = target }
        if !text.isEmpty { obj["text"] = text }
        return ChatOut(text: reply, phone: [obj])
    }

    private static func match(_ pattern: String, _ text: String) -> String? {
        guard let re = try? NSRegularExpression(pattern: pattern) else { return nil }
        let range = NSRange(text.startIndex..., in: text)
        guard let m = re.firstMatch(in: text, range: range), m.numberOfRanges > 1,
              let r = Range(m.range(at: 1), in: text)
        else { return nil }
        if m.numberOfRanges > 2, let r2 = Range(m.range(at: 2), in: text) {
            return "\(text[r])|\(text[r2])"
        }
        return String(text[r])
    }
}

File: android/app/src/main/java/app/gsuss/asistente/PhoneLocal.kt
package app.gsuss.asistente

import android.content.Context
import org.json.JSONObject
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * Runs on the phone without the PC. PhoneHands still executes the intents.
 * Mirrors everyday local commands from jarvis/local.py for offline/mobile use.
 */
object PhoneLocal {
    private val openRe = Regex(
        """(?:abr[ií]|abrime|abrir|abre|open|lanz[aá]|ejecut[aá]|and[aá]\s+a|quiero\s+que\s+abras?|sac[aá])\s+(?:la\s+|el\s+|app\s+(?:de\s+)?)?(.+)$""",
        RegexOption.IGNORE_CASE,
    )
    private val noteRe = Regex(
        """^(?:anot[aá]|nota[:\s]+|record[aá]\s+esto[:\s]*|tom[aá]\s+nota(?:\s+de(?:\s+que)?)?)\s*(.+)$""",
        setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL),
    )
    private val rememberRe = Regex(
        """^(?:acordate(?:\s+que)?|record[aá]\s+que)\s+(.+?)\s+(?:es|=|:)\s+(.+)$""",
        setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL),
    )
    private val callRe = Regex(
        """(?:llam[aá]|marca[lr]?|disc[aá])\s+(?:al\s+|a\s+)?([+\d][\d\s\-()]{6,})""",
        RegexOption.IGNORE_CASE,
    )
    private val smsRe = Regex(
        """(?:sms|mensaje(?:\s+de\s+texto)?)\s+(?:a|al)\s+([+\d][\d\s\-()]{6,})\s*[:\-]?\s*(.*)$""",
        setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL),
    )
    private val waRe = Regex(
        """(?:whats?app|wpp|wasap)\s+(?:a|al|para)?\s*([+\d][\d\s\-]{7,18})\s*[:\-]\s*(.+)$""",
        setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL),
    )
    private val mapsRe = Regex(
        """^(?:c[oó]mo\s+llego(?:\s+a)?|mapas?|ruta(?:\s+a)?|llevame\s+a|ll[eé]vame\s+a)\s+(.+)$""",
        RegexOption.IGNORE_CASE,
    )
    private val musicRe = Regex(
        """(?:poneme|pon[eé]|reproduc[ií]|reproducir|escuchar|play|tirame)\s+""" +
            """(?:a\s+reproducir\s+)?""" +
            """(?:un\s+tema\s+de\s+|una\s+canci[oó]n\s+(?:de\s+)?|la\s+canci[oó]n\s+(?:de\s+)?|""" +
            """el\s+tema\s+(?:de\s+)?|m[uú]sica\s+(?:de\s+)?)?""" +
            """(.+?)""" +
            """(?:\s+en\s+(?:el\s+)?(spotify|youtube))?\s*$""",
        RegexOption.IGNORE_CASE,
    )
    private val timerRe = Regex(
        """(?:timer|temporizador|avis[aá]me|record[aá]me)\s+(?:en\s+)?(\d+)\s*(min|mins|minuto|minutos|hora|horas)\b(?:\s+(?:para|que|:|de)\s*(.+))?""",
        RegexOption.IGNORE_CASE,
    )
    private val alarmRe = Regex(
        """\b(?:alarma|alarm)\s+(?:a\s+las\s+|para\s+las\s+|a\s+)?(\d{1,2})(?:[:\.](\d{2}))?\b""",
        RegexOption.IGNORE_CASE,
    )
    private val bareApps = setOf(
        "spotify", "whatsapp", "telegram", "instagram", "youtube", "tiktok",
        "chrome", "gmail", "maps", "netflix", "discord", "twitch",
    )

    fun handle(
        context: Context,
        raw: String,
        notes: NotesCache,
        prefs: Prefs,
    ): ChatOut? {
        val text = raw.replace(Regex("""^(?:hey\s+)?ilaria\b[\s,.:\-]*""", RegexOption.IGNORE_CASE), "").trim()
        if (text.isBlank()) return ChatOut("Te escucho.")
        val lower = text.lowercase(Locale.getDefault())

        if (Regex("""\b(ayuda|help|qu[eé] pod[eé]s|comandos)\b""").containsMatchIn(lower)) {
            return ChatOut(
                "En el celular: apps (no bancarias), Spotify/YouTube, linterna, cámara, mapas, " +
                    "llamadas, SMS/WhatsApp borrador, volumen, alarma/timer, wifi/bluetooth, " +
                    "portapapeles, traducir, notas. Con la PC en Wi‑Fi, pienso con Ilaria de ahí.",
            )
        }
        if (Regex("""\b(hora|fecha|qu[eé] d[ií]a)\b""").containsMatchIn(lower) || lower in setOf("ahora", "hora", "fecha")) {
            val zone = try {
                ZoneId.of("America/Argentina/Buenos_Aires")
            } catch (_: Exception) {
                ZoneId.systemDefault()
            }
            val now = ZonedDateTime.now(zone)
            val fmt = DateTimeFormatter.ofPattern("EEEE d 'de' MMMM, HH:mm", Locale("es", "AR"))
            return ChatOut(now.format(fmt))
        }
        if (Regex("""\b(linterna|flashlight|torch)\b""").containsMatchIn(lower)) {
            val off = Regex("""\b(apag|off|sac[aá])\b""").containsMatchIn(lower)
            return hands("torch", if (off) "off" else "on", "Listo.")
        }
        if (Regex("""\b(c[aá]mara|camera)\b""").containsMatchIn(lower) &&
            Regex("""\b(abr|sac[aá]|pon|tir[aá])\b""").containsMatchIn(lower)
        ) {
            return hands("camera", reply = "Cámara.")
        }
        if (Regex("""\b(galer[ií]a|fotos|photos)\b""").containsMatchIn(lower) &&
            Regex("""\b(abr|mostr)\b""").containsMatchIn(lower)
        ) {
            return hands("gallery", reply = "Galería.")
        }
        if (Regex("""\b(wifi|wi[\-\s]?fi)\b""").containsMatchIn(lower) &&
            Regex("""\b(abr|ajustes|config)\b""").containsMatchIn(lower)
        ) {
            return hands("open_wifi_settings", reply = "Wi‑Fi.")
        }
        if (Regex("""\b(bluetooth)\b""").containsMatchIn(lower)) {
            return hands("bluetooth", reply = "Bluetooth.")
        }
        if (Regex("""\b(ajustes|configuraci[oó]n|settings)\b""").containsMatchIn(lower)) {
            return hands("settings", reply = "Ajustes.")
        }
        if (Regex("""\b(silenci(?:ar|[oaá])|mute)\b""").containsMatchIn(lower)) {
            return hands("volume", "mute", "Silencio.")
        }
        val volN = Regex("""(?:volumen|volume)(?:\s+(?:al|a|en))?\s+(\d{1,3})""").find(lower)
        if (volN != null) return hands("volume", volN.groupValues[1], "Volumen ${volN.groupValues[1]}%.")
        if (Regex("""\b(sub[ií]|m[aá]s).{0,12}volumen|volumen.{0,12}(sub[ií]|m[aá]s)\b""").containsMatchIn(lower)) {
            return hands("volume", "up", "Subo volumen.")
        }
        if (Regex("""\b(baj[aá]|menos).{0,12}volumen|volumen.{0,12}(baj[aá]|menos)\b""").containsMatchIn(lower)) {
            return hands("volume", "down", "Bajo volumen.")
        }
        if (Regex("""\b(captura|screenshot)\b""").containsMatchIn(lower)) {
            return hands("screenshot", reply = "Power + Volumen abajo para capturar.")
        }
        if (Regex("""\b(bloque[aá]|lock).{0,12}(pantalla|tel[eé]fono|celular)\b""").containsMatchIn(lower)) {
            return hands("lock", reply = "Bloqueo.")
        }
        val clipSet = Regex("""^(?:copi[aá]|copiar)\s*[:\-]?\s+(.+)$""", setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL)).find(text)
        if (clipSet != null) {
            return hands("clipboard", text = clipSet.groupValues[1].trim(), reply = "Copiado.")
        }
        if (Regex("""\b(portapapeles|qu[eé]\s+copi[eé])\b""").containsMatchIn(lower)) {
            return hands("clipboard_get", reply = "Leo el portapapeles.")
        }
        val translate = Regex("""^(?:traduc[ií]|traducir|translate)\s+(.+)$""", RegexOption.IGNORE_CASE).find(text)
        if (translate != null) {
            return hands("translate", text = translate.groupValues[1].trim(), reply = "Traduzco.")
        }
        val yt = Regex("""^(?:busc[aá]|busca)\s+(?:en\s+)?youtube\s+(.+)$""", RegexOption.IGNORE_CASE).find(text)
        if (yt != null) return hands("youtube", yt.groupValues[1].trim(), "YouTube.")

        val alarm = alarmRe.find(lower)
        if (alarm != null) {
            val hh = alarm.groupValues[1]
            val mm = alarm.groupValues.getOrNull(2)?.ifBlank { "00" } ?: "00"
            return hands("alarm", "$hh:$mm", "Alarma $hh:$mm.")
        }
        val timer = timerRe.find(lower)
        if (timer != null) {
            var mins = timer.groupValues[1].toIntOrNull() ?: 5
            if (timer.groupValues[2].startsWith("hora")) mins *= 60
            return hands("timer", mins.toString(), "Timer $mins min.")
        }

        val call = callRe.find(text)
        if (call != null) return hands("call", call.groupValues[1].filter { it.isDigit() || it == '+' }, "Abro el marcador.")
        val sms = smsRe.find(text)
        if (sms != null) {
            return hands(
                "sms",
                sms.groupValues[1].filter { it.isDigit() || it == '+' },
                text = sms.groupValues.getOrNull(2)?.trim().orEmpty(),
                reply = "Borrador SMS.",
            )
        }
        val wa = waRe.find(text)
        if (wa != null) {
            return hands(
                "whatsapp",
                wa.groupValues[1].filter { it.isDigit() },
                text = wa.groupValues[2].trim(),
                reply = "WhatsApp borrador.",
            )
        }
        val maps = mapsRe.find(text)
        if (maps != null) return hands("navigate", maps.groupValues[1].trim(), "Te armo la ruta.")

        if (Regex("""\b(contactos)\b""").containsMatchIn(lower)) return hands("contacts", reply = "Contactos.")
        if (Regex("""\b(calendario|agenda)\b""").containsMatchIn(lower)) return hands("calendar", reply = "Calendario.")

        val music = musicRe.find(text)
        if (music != null && !Regex("""\b(volumen|timer|alarma|linterna|recordatorio)\b""").containsMatchIn(lower)) {
            var query = music.groupValues[1].trim().trim('.', '!', '?')
            query = query.replace(Regex("""\s+en\s+(el\s+)?(spotify|youtube)$""", RegexOption.IGNORE_CASE), "").trim()
            query = query.replace(Regex("""^(spotify|youtube)\s+""", RegexOption.IGNORE_CASE), "").trim()
            val platform = music.groupValues.getOrNull(2)?.lowercase(Locale.getDefault()).orEmpty()
            val apps = setOf("spotify", "youtube", "whatsapp", "telegram", "instagram", "maps", "gmail", "chrome", "tiktok")
            when {
                query in setOf("spotify", "youtube", "musica", "música") -> {
                    val target = if ("youtube" in query) "youtube" else "spotify"
                    return hands("open_app", target, "Abro $target.")
                }
                query.isNotBlank() && query !in apps -> {
                    return if (platform == "youtube" || "youtube" in lower) {
                        hands("youtube", query, "Busco $query en YouTube.")
                    } else {
                        hands("music", query, "Busco $query en Spotify.")
                    }
                }
            }
        }

        val remembered = rememberRe.find(text)
        if (remembered != null) {
            prefs.putFact(remembered.groupValues[1].trim(), remembered.groupValues[2].trim())
            return ChatOut("Anotado.")
        }
        if (Regex("""\b(qu[eé] sab[eé]s|mis datos|memoria)\b""").containsMatchIn(lower)) {
            val facts = prefs.facts()
            val body = if (facts.isEmpty()) "Todavía no me dijiste datos." else facts.entries.joinToString("\n") { "- ${it.key}: ${it.value}" }
            return ChatOut(body)
        }
        val note = noteRe.find(text)
        if (note != null) {
            notes.addLocal(note.groupValues[1].trim())
            return ChatOut("Nota guardada en el celular. Se sincroniza con la PC cuando hay enlace.")
        }
        if (Regex("""^(notas|mis notas|lista(r)? notas)$""").containsMatchIn(lower)) {
            val lines = notes.snapshot().texts()
            return ChatOut(if (lines.isEmpty()) "No hay notas." else lines.mapIndexed { i, t -> "${i + 1}. $t" }.joinToString("\n"))
        }
        if (text.startsWith("http://", true) || text.startsWith("https://", true)) {
            return hands("browser", text.trim(), "Abro el link.")
        }
        val opened = openRe.find(text)
        if (opened != null) {
            val target = opened.groupValues[1].trim().trim('.', '!', '?')
            if (BankApps.blocked(target)) return ChatOut("No abro apps bancarias.")
            when (target.lowercase(Locale.getDefault())) {
                "youtube", "yt" -> return hands("browser", "https://www.youtube.com", "YouTube.")
                "gmail", "correo", "mail" -> return hands("browser", "https://mail.google.com", "Gmail.")
                "maps", "mapas" -> return hands("maps", "acá", "Mapas.")
                "netflix" -> return hands("browser", "https://www.netflix.com", "Netflix.")
            }
            return hands("open_app", target, "Abro $target.")
        }
        val bare = bareApps.firstOrNull { it == lower.trim() }
        if (bare != null) return hands("open_app", bare, "Abro $bare.")
        return null
    }

    fun fallback(prefs: Prefs, linked: Boolean): String {
        val who = prefs.displayName.ifBlank { prefs.username }.ifBlank { "vos" }
        return if (linked) {
            "No pude completar eso en el celular. Probá de nuevo."
        } else {
            "Estoy en el celular, $who — independiente de la PC. Apps, notas, volumen y linterna van acá. " +
                "Si querés, en Perfil sincronizás con la PC (mismo Wi‑Fi)."
        }
    }

    private fun hands(
        action: String,
        target: String = "",
        reply: String,
        text: String = "",
    ): ChatOut {
        val obj = JSONObject().put("action", action)
        if (target.isNotBlank()) obj.put("target", target)
        if (text.isNotBlank()) obj.put("text", text)
        return ChatOut(reply, listOf(obj))
    }
}
